Consumer closes with an exception NoSuchElementException

Use case: 
Poller is configured and subscribed for 2 topics.
After polling if message needs further processing then push to separate topic.
Polling for messages 60>k messages per hour.

Version: 0.5.2.2

**Configuration:**
Order by Partitions and max concurrency is 12.

Consumer Configuration:
```
props.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
props.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest");

```

During the high load of messages > 60k per hour I have seen that the consumer closes down and there are following exceptions seen

I think the PartitionState.getOffsetHighestSequentialSucceeded() method implementation can be changed to follow as the current implementation is not thread safe and have possibility of failing with NoSuchElementException.

**Current Implementation:**
```
public long getOffsetHighestSequentialSucceeded() {
 if (this.incompleteOffsets.isEmpty()) {
            return this.offsetHighestSeen;
        } else {
            return this.incompleteOffsets.first() - 1;
        }
}
```

**Possible New Implementation:**
```
public long getOffsetHighestSequentialSucceeded() {
        Long offset = this.incompleteOffsets.ceiling(0l);
        return offset==null ? offsetHighestSeen : offset-1;
}

```
If the implementation map is empty this will return null and exception can be avoided.


**Exceptions:**


"message":"Unknown error","logger_name":"io.confluent.parallelconsumer.internal.BrokerPollSystem","thread_name":"pc-broker-poll","level":"ERROR","level_value":40000,"stack_trace":"java.util.NoSuchElementException: null\n\tat java.base/java.util.concurrent.ConcurrentSkipListMap.firstKey(Unknown Source)\n\tat java.base/java.util.concurrent.ConcurrentSkipListSet.first(Unknown Source)\n\tat io.confluent.parallelconsumer.state.PartitionState.getOffsetHighestSequentialSucceeded(PartitionState.java:267)\n\tat io.confluent.parallelconsumer.state.PartitionState.getNextExpectedPolledOffset(PartitionState.java:235)\n\tat io.confluent.parallelconsumer.state.PartitionState.createOffsetAndMetadata(PartitionState.java:225)\n\tat io.confluent.parallelconsumer.state.PartitionState.getCommitDataIfDirty(PartitionState.java:218)\n\tat io.confluent.parallelconsumer.state.PartitionStateManager.collectDirtyCommitData(PartitionStateManager.java:373)\n\tat io.confluent.parallelconsumer.state.WorkManager.collectCommitDataForDirtyPartitions(WorkManager.java:198)\n\tat io.confluent.parallelconsumer.internal.AbstractOffsetCommitter.retrieveOffsetsAndCommit(AbstractOffsetCommitter.java:31)\n\tat io.confluent.parallelconsumer.internal.ConsumerOffsetCommitter.maybeDoCommit(ConsumerOffsetCommitter.java:179)\n\tat java.base/java.util.Optional.ifPresent(Unknown Source)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.maybeDoCommit(BrokerPollSystem.java:323)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.controlLoop(BrokerPollSystem.java:118)\n\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\n

Error from poll control thread, will attempt controlled shutdown, then rethrow. Error: Error in BrokerPollSystem system.","logger_name":"io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor","thread_name":"pc-control","level":"ERROR","level_value":40000,"stack_trace":"io.confluent.parallelconsumer.internal.InternalRuntimeError: Error in BrokerPollSystem system.\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.supervise(BrokerPollSystem.java:100)\n\tat io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor.controlLoop(AbstractParallelEoSStreamProcessor.java:703)\n\tat io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor.lambda$supervisorLoop$5(AbstractParallelEoSStreamProcessor.java:639)\n\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.util.concurrent.ExecutionException: java.util.NoSuchElementException\n\tat java.base/java.util.concurrent.FutureTask.report(Unknown Source)\n\tat java.base/java.util.concurrent.FutureTask.get(Unknown Source)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.supervise(BrokerPollSystem.java:98)\n\t... 6 common frames omitted\nCaused by: java.util.NoSuchElementException: null\n\tat java.base/java.util.concurrent.ConcurrentSkipListMap.firstKey(Unknown Source)\n\tat java.base/java.util.concurrent.ConcurrentSkipListSet.first(Unknown Source)\n\tat io.confluent.parallelconsumer.state.PartitionState.getOffsetHighestSequentialSucceeded(PartitionState.java:267)\n\tat io.confluent.parallelconsumer.state.PartitionState.getNextExpectedPolledOffset(PartitionState.java:235)\n\tat io.confluent.parallelconsumer.state.PartitionState.createOffsetAndMetadata(PartitionState.java:225)\n\tat io.confluent.parallelconsumer.state.PartitionState.getCommitDataIfDirty(PartitionState.java:218)\n\tat io.confluent.parallelconsumer.state.PartitionStateManager.collectDirtyCommitData(PartitionStateManager.java:373)\n\tat io.confluent.parallelconsumer.state.WorkManager.collectCommitDataForDirtyPartitions(WorkManager.java:198)\n\tat io.confluent.parallelconsumer.internal.AbstractOffsetCommitter.retrieveOffsetsAndCommit(AbstractOffsetCommitter.java:31)\n\tat io.confluent.parallelconsumer.internal.ConsumerOffsetCommitter.maybeDoCommit(ConsumerOffsetCommitter.java:179)\n\tat java.base/java.util.Optional.ifPresent(Unknown Source)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.maybeDoCommit(BrokerPollSystem.java:323)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.controlLoop(BrokerPollSystem.java:118)\n\t... 4 common frames omitted\n","dd.service":"uc3-connector","dd.version":"0.0.58-4"}

message":"Execution or timeout exception waiting for broker poller thread to finish","logger_name":"io.confluent.parallelconsumer.internal.BrokerPollSystem","thread_name":"pc-control","level":"ERROR","level_value":40000,"stack_trace":"java.util.concurrent.ExecutionException: java.util.NoSuchElementException\n\tat java.base/java.util.concurrent.FutureTask.report(Unknown Source)\n\tat java.base/java.util.concurrent.FutureTask.get(Unknown Source)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.closeAndWait(BrokerPollSystem.java:244)\n\tat io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor.doClose(AbstractParallelEoSStreamProcessor.java:515)\n\tat io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor.lambda$supervisorLoop$5(AbstractParallelEoSStreamProcessor.java:642)\n\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\nCaused by: java.util.NoSuchElementException: null\n\tat java.base/java.util.concurrent.ConcurrentSkipListMap.firstKey(Unknown Source)\n\tat java.base/java.util.concurrent.ConcurrentSkipListSet.first(Unknown Source)\n\tat io.confluent.parallelconsumer.state.PartitionState.getOffsetHighestSequentialSucceeded(PartitionState.java:267)\n\tat io.confluent.parallelconsumer.state.PartitionState.getNextExpectedPolledOffset(PartitionState.java:235)\n\tat io.confluent.parallelconsumer.state.PartitionState.createOffsetAndMetadata(PartitionState.java:225)\n\tat io.confluent.parallelconsumer.state.PartitionState.getCommitDataIfDirty(PartitionState.java:218)\n\tat io.confluent.parallelconsumer.state.PartitionStateManager.collectDirtyCommitData(PartitionStateManager.java:373)\n\tat io.confluent.parallelconsumer.state.WorkManager.collectCommitDataForDirtyPartitions(WorkManager.java:198)\n\tat io.confluent.parallelconsumer.internal.AbstractOffsetCommitter.retrieveOffsetsAndCommit(AbstractOffsetCommitter.java:31)\n\tat io.confluent.parallelconsumer.internal.ConsumerOffsetCommitter.maybeDoCommit(ConsumerOffsetCommitter.java:179)\n\tat java.base/java.util.Optional.ifPresent(Unknown Source)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.maybeDoCommit(BrokerPollSystem.java:323)\n\tat io.confluent.parallelconsumer.internal.BrokerPollSystem.controlLoop(BrokerPollSystem.java:118)\n\t... 4 common frames omitted

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -23,6 +23,9 @@ endif::[]
 ** Fixes a potential race condition which could cause offset leaks between transactions boundaries.
 ** Introduces lock acquisition timeouts.
 ** Fixes a potential issue with removing records from the retry queue incorrectly, by having an inconsistency between compareTo and equals in the retry TreeMap.
+
+=== Fixes
+
 * fixes #419 NoSuchElementException during race condition in PartitionState (#422)
 
 == v0.5.2.3

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -23,6 +23,7 @@ endif::[]
 ** Fixes a potential race condition which could cause offset leaks between transactions boundaries.
 ** Introduces lock acquisition timeouts.
 ** Fixes a potential issue with removing records from the retry queue incorrectly, by having an inconsistency between compareTo and equals in the retry TreeMap.
+* fixes #419 NoSuchElementException during race condition in PartitionState (#422)
 
 == v0.5.2.3
 

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -23,17 +23,22 @@ endif::[]
 ** Fixes a potential race condition which could cause offset leaks between transactions boundaries.
 ** Introduces lock acquisition timeouts.
 ** Fixes a potential issue with removing records from the retry queue incorrectly, by having an inconsistency between compareTo and equals in the retry TreeMap.
+* Adds a very simple Dependency Injection system modeled on Dagger (#398)
+* Various refactorings e.g. new ProducerWrap
+
+* Dependencies
+** build(deps): prod: zstd, reactor, dev: podam, progressbar, postgresql maven-plugins: versions, help (#420)
+** build(deps-dev): bump postgresql from 42.4.1 to 42.5.0
+** bump podam, progressbar, zstd, reactor
+** build(deps): bump versions-maven-plugin from 2.11.0 to 2.12.0
+** build(deps): bump maven-help-plugin from 3.2.0 to 3.3.0
+** build(deps-dev): bump Confluent Platform Kafka Broker to 7.2.2 (#421)
+** build(deps): Upgrade to AK 3.3.0 (#309)
 
 === Fixes
 
 * fixes #419 NoSuchElementException during race condition in PartitionState (#422)
 
-== v0.5.2.3
-
-=== Improvements
-
-- Adds a very simple Dependency Injection system modeled on Dagger (#398)
-- Various refactorings e.g. new ProducerWrap
 
 == v0.5.2.2
 

--- a/README.adoc
+++ b/README.adoc
@@ -1279,6 +1279,7 @@ endif::[]
 ** Fixes a potential race condition which could cause offset leaks between transactions boundaries.
 ** Introduces lock acquisition timeouts.
 ** Fixes a potential issue with removing records from the retry queue incorrectly, by having an inconsistency between compareTo and equals in the retry TreeMap.
+* fixes #419 NoSuchElementException during race condition in PartitionState (#422)
 
 == v0.5.2.3
 

--- a/parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/PartitionState.java
+++ b/parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/PartitionState.java
@@ -261,17 +261,30 @@ public class PartitionState<K, V> {
      * lowest incomplete offset.
      */
     public long getOffsetHighestSequentialSucceeded() {
-        if (this.incompleteOffsets.isEmpty()) {
-            return this.offsetHighestSeen;
+        /*
+         * Capture the current value in case it's changed during this operation - because if more records are added to
+         * the queue, after looking at the incompleteOffsets, offsetHighestSeen could increase drastically and will be
+         * incorrect for the value of getOffsetHighestSequentialSucceeded. So this is a ~pessimistic solution - as in a
+         * race case, there may be a higher getOffsetHighestSequentialSucceeded from the incompleteOffsets collection,
+         * but it will always at lease be pessimistically correct in terms of committing offsets to the broker.
+         *
+         * See #200 for the complete correct solution.
+         */
+        long currentOffsetHighestSeen = offsetHighestSeen;
+        Long firstIncompleteOffset = incompleteOffsets.ceiling(KAFKA_OFFSET_ABSENCE);
+        boolean incompleteOffsetsWasEmpty = firstIncompleteOffset == null;
+
+        if (incompleteOffsetsWasEmpty) {
+            return currentOffsetHighestSeen;
         } else {
-            return this.incompleteOffsets.first() - 1;
+            return firstIncompleteOffset - 1;
         }
     }
 
     /**
      * Tries to encode the incomplete offsets for this partition. This may not be possible if there are none, or if no
-     * encodings are possible ({@link NoEncodingPossibleException}. Encoding may not be possible of - see {@link
-     * OffsetMapCodecManager#makeOffsetMetadataPayload}.
+     * encodings are possible ({@link NoEncodingPossibleException}. Encoding may not be possible of - see
+     * {@link OffsetMapCodecManager#makeOffsetMetadataPayload}.
      *
      * @return if possible, the String encoded offset map
      */

